api/v1beta1/gcpmachine_types.go

```diff
@@ -146,6 +149,32 @@ const (
 	HostMaintenancePolicyTerminate HostMaintenancePolicy = "Terminate"
 )
 
+// CustomerEncryptionKey supports both CMEKs via the KmsKeyName or CSEKs via the RawKey or RsaEncryptedKey.
+type CustomerEncryptionKey struct {
+	// The name of the encryption key that is stored in Google Cloud KMS. For example:
+	// "kmsKeyName": "projects/kms_project_id/locations/region/keyRings/key_region/cryptoKeys/key
+	KmsKeyName string `json:"kmsKeyName,omitempty"`
+	// The service account being used for the encryption request for the given KMS key.
+	// If absent, the Compute Engine default service account is used. For example:
+	// "kmsKeyServiceAccount": "name@project_id.iam.gserviceaccount.com/
+	KmsKeyServiceAccount string `json:"kmsKeyServiceAccount,omitempty"`
+	// Specifies a 256-bit customer-supplied encryption key, encoded in RFC 4648
+	// base64 to either encrypt or decrypt this resource. You can provide either the rawKey or the rsaEncryptedKey.
+	// For example: "rawKey": "SGVsbG8gZnJvbSBHb29nbGUgQ2xvdWQgUGxhdGZvcm0="
+	RawKey string `json:"rawKey,omitempty"`
+	// Specifies an RFC 4648 base64 encoded, RSA-wrapped 2048-bit customer-supplied encryption
+	// key to either encrypt or decrypt this resource. You can provide either the rawKey or the
+	// rsaEncryptedKey.
+	// For example: "rsaEncryptedKey": "ieCx/NcW06PcT7Ep1X6LUTc/hLvUDYyzSZPPVCVPTVEohpeHASqC8uw5TzyO9U+Fka9JFHi
+	// z0mBibXUInrC/jEk014kCK/NPjYgEMOyssZ4ZINPKxlUh2zn1bV+MCaTICrdmuSBTWlUUiFoDi
+	// D6PYznLwh8ZNdaheCeZ8ewEXgFQ8V+sDroLaN3Xs3MDTXQEMMoNUXMCZEIpg9Vtp9x2oe=="
+	// The key must meet the following requirements before you can provide it to Compute Engine:
+	// 1. The key is wrapped using a RSA public key certificate provided by Google.
+	// 2. After being wrapped, the key must be encoded in RFC 4648 base64 encoding.
+	// Gets the RSA public key certificate provided by Google at: https://cloud-certs.storage.googleapis.com/google-cloud-csek-ingress.pem
+	RsaEncryptedKey string `json:"rsaEncryptedKey,omitempty"`
```

Comment:
Which of these fields work together, and which do not? Typically in Kube when there's a choice, either supply a reference to a key, or provide the raw key, we use a discriminated union and make the user choose.

Reply:
> Which of these fields work together, and which do not? Typically in Kube when there's a choice, either supply a reference to a key, or provide the raw key, we use a discriminated union and make the user choose.

The way I understand it - the encryption can be defined by only one of KmsKeyName, RawKey, or RsaEncryptedKey. KmsKeyServiceAccount is applicable to any of the encryption choices. 
Is there an example of a way to represent this?

Reply:
Thank you, will update.

Reply:
Added a KeyType and separate structs to distinguish the settings for CustomerManaged or CustomerSupplied keys. I didn't further split out the RawKey from the RsaEncryptedKey as its documented that either can be provided.